src/http.ts

```diff
@@ -30,6 +39,130 @@ export interface SourceImageOptions {
   requestEtag?: string
 }
 
+interface UsageTrackingItem {
+  runningCount: number;
+  cacheKey: string;
+  lastAccess: number;
+  inputCacheFile: string
+  size: number
+}
+
+type UsageTracking = Record<
+  string,
+  UsageTrackingItem
+>;
+
+const USAGE_TRACKING_KEY = 'usage-tracking'
+const trackingLock = new Lock()
+
+async function getTracking (metadataStore: Storage): Promise<UsageTracking> {
+  return ((await metadataStore.getItem(USAGE_TRACKING_KEY)) ??
+    {}) as UsageTracking
+}
+
+async function markUsageStart (
+  metadataStore: Storage,
+  cacheKey: string,
+  inputCacheFile: string
+): Promise<void> {
+  await trackingLock.acquire()
+  try {
+    const tracking = await getTracking(metadataStore)
+
+    let usageTrackingItem = tracking[cacheKey]
+    if (!usageTrackingItem) {
+      tracking[cacheKey] = usageTrackingItem = {
+        runningCount: 0,
+        lastAccess: 0,
+        size: 0,
+        cacheKey,
+        inputCacheFile
+      }
+    }
+
+    usageTrackingItem.runningCount++
+    usageTrackingItem.lastAccess = Date.now()
+
+    await metadataStore.setItem(USAGE_TRACKING_KEY, tracking)
+  } finally {
+    trackingLock.release()
+  }
+}
+
+async function markUsageComplete (
+  metadataStore: Storage,
+  cacheKey: string
+
+) {
+  await trackingLock.acquire()
+  try {
+    const tracking = await getTracking(metadataStore)
+
+    const usageTrackingItem = tracking[cacheKey]
+    if (usageTrackingItem) {
+      usageTrackingItem.runningCount--
+
+      if (await pathExists(usageTrackingItem.inputCacheFile)) {
+        const { size } = await stat(usageTrackingItem.inputCacheFile)
+        usageTrackingItem.size = size
+      } else {
+        // If the file doesn't exist, we can't track it
+        delete tracking[cacheKey]
+      }
+
+      await metadataStore.setItem(USAGE_TRACKING_KEY, tracking)
+    }
+  } finally {
+    trackingLock.release()
+  }
+}
+
+export const CACHE_PRUNING_THRESHOLD = 50 * 1024 * 1024
```

Comment:
Cache pruning right now only happens before source image download start - before doing any work we check if we are over the threshold first and then try to get below that threshold and only after that we actually start to download source image.

This will allow image of any size to be processed (as long as there enough disk space). It also mean that that image will likely be one that gets pruned quickly when another request comes in (as long as all image transformation that need that image already finished)

This is why I mentioned this in description:

> Above doesn't guarantee not running into disk space issue and even more so if same lambdas can handle multiple requests concurrently (not sure what's the behavior here), but it should at least reduce possibility of running into free disk issues.

This is also why threshold is set to be so low (AFAIK min disk space for lambdas is 512MB, so 50MB threshold is very conservative)

Also side note - if input image is over 50MB I suspect we either hit timeout due to amount of CPU cycles needed to transform such huge image or we hit memory issues when sharp / vips will try to decompress such image into raw image bitmap.